template/DEVELOPING/environment-variables.md

```diff
@@ -1,11 +1,54 @@
 # Environment Variables
 
-We use
-[@nest/env](https://github.com/vercel/next.js/blob/master/docs/basic-features/environment-variables.md#loading-environment-variables)
-to load environment variables from `.env*` files in both the frontend and
-backend packages. See the @nest/env readme for detailed documentation.
+## Reading Environment Variables
 
-Briefly you can have multiple env files that are loaded depending
+All of the environment variables that the app reads are declared in [Env][]. We
+use [class-validator][] to specify whether each variable is required, to provide
+default values, and to specify any additional validation. For example,
+
+[env]: ../packages/backend/src/env.ts
+[class-validator]: https://github.com/typestack/class-validator#readme
+
+```ts
+export class Env {
+  // built-in validator checks that the variable is formatted as a URL
+  @IsUrl({
+    require_protocol: true,
+    require_valid_protocol: false,
+    require_tld: false,
+  })
+  DATABASE_URL!: string
+  // ^ no `@IsOptional()` means this one is required
+
+  @IsPort() // built-in validator checks value is a number from 0-65535
+  @IsOptional() // this variable is optional
+  PORT = "3000" // default value
+}
+```
+
+> ℹ️ We know that required variables in `Env` will be set thanks to validation
+> from class-validator. TypeScript doesn't know that, so we put an `!` after
+> class property names to assert non-nullability.
+
+If you want to introduce a new environment variable please add it to `Env` -
+avoid reading directly from `process.env`. That way we can look in one place to
+see all of the environment variables that the app uses.
+
+There is one instance of `Env` which is created by [ConfigService][]. And in
```

Comment:
Do you mean Env is a singleton? Otherwise I might be confused that multiple Envs might need to be created.